src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
+            Node<T> newTail = new Node<T>(tail, value, null);
+            newTail.prev.next = newTail;
+            tail = newTail;
+            size++;
+        } else {
+            Node<T> valueToAdd = new Node<>(elementFinder(index).prev, value, elementFinder(index));
+            valueToAdd.prev.next = valueToAdd;
+            valueToAdd.next.prev = valueToAdd;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementFinder(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = elementFinder(index).item;
+        elementFinder(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue = elementFinder(index).item;
+        unlink(elementFinder(index));
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> i = head; i != null; i = i.next) {
+                if (object == i.item) {
+                    unlink(i);
+                    return true;
+                }
+            }
+            return false;
+        }
+        for (Node<T> i = head; i != null; i = i.next) {
+            if (object.equals(i.item)) {
+                unlink(i);
+                return true;
+            }
+        }
         return false;
     }
 
+    private T unlink(Node<T> nodeToUnlink) {
+        final T unlinkedElement = nodeToUnlink.item;
+        Node<T> previos = nodeToUnlink.prev;
+        Node<T> next = nodeToUnlink.next;
+        if (previos == null) {
+            head = next;
+        } else {
+            previos.next = next;
+            nodeToUnlink.prev = null;
+        }
+        if (next == null) {
+            tail = previos;
+        } else {
+            next.prev = previos;
+            nodeToUnlink.next = null;
+        }
+        nodeToUnlink.item = null;
+        size--;
+        return unlinkedElement;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("can not reach element with this index");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("can not reach element with this index");
+        }
+    }
+
+    private Node<T> elementFinder(int index) {
```

Comment:
```suggestion
    private Node<T> findNodeByIndex(int index) {
```
method names must be verbs